### PR TITLE
ensure backward compatibility for Web3 constructor

### PR DESCRIPTION
**Pull request by Muhammad-Altabba**
## Description
Fixes [#(5196)](https://github.com/ChainSafe/web3.js/issues/5196)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [x] I ran `yarn run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

File: packages/web3/src/index.ts [hunk 1]
```diff
@@ -327,3 +327,8 @@ export * from './web3';
 export * from './types';
 
 export default Web3;
+
+// The following JavaScript command is used instead of the equivalent TypeScript `export = Web3`, 
+//  to allow exporting the types along with Web3 constructor as the default function,
+//  in order to maintain backward compatibility.
+module.exports = Web3;
```
Comment (nazarhussain) [inline packages/web3/src/index.ts]: Need to disable the TS error. 

```suggestion
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
module.exports = Web3;
```

**Review by nazarhussain — DISMISSED** [review]
Please make this change to every `default` export in all the packages.

File: packages/web3/src/index.ts (hunk 1 above)
Comment (Muhammad-Altabba) [reply packages/web3/src/index.ts]: Thanks for the hint. However, I changed the approach :blush:

**Comment by Muhammad-Altabba** [comment]
With the last commit, the `export` and `import` behaves exactly as it was in version 1.x (this is as far as I know. And let me know if this is not true) :smile: 

However, this also means that the `types` inside `web3` package are no longer exported. So, I recommend having a follow-up task to put those types possibly at `web3-core`. This can be done perhaps in `beta` or at a later version. And this could be done along with the refactoring proposed at https://github.com/ChainSafe/web3.js/discussions/5223, or separately.

File: packages/web3/src/index.ts [hunk 2]
```diff
@@ -323,7 +323,4 @@ along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
  */
 import { Web3 } from './web3';
 
-export * from './web3';
-export * from './types';
```
Comment (nazarhussain) [inline packages/web3/src/index.ts:327]: We can't skip exporting types from this package as well the named export.

**Review by nazarhussain — CHANGES_REQUESTED** [review]
Apart form the mentioned change, we need to apply same changes to all packages which are using `export default`.

File: packages/web3/src/index.ts (hunk 2 above)
Comment (luu-alex) [reply packages/web3/src/index.ts:327]: why can't we skip exporting types? using muhammads branch in a test project and looks like it still gets type safety for `Web3EthInterface`

**Review by nazarhussain — APPROVED** [review]

File: packages/web3/src/index.ts (hunk 2 above)
Comment (nazarhussain) [reply packages/web3/src/index.ts:327]: Unfortunately we have to skip the types export due to a limitation in TS. We can't combine name and default exports in CommonJS.

https://github.com/microsoft/TypeScript/issues/2719#issuecomment-123960556

**Review by jdevcs — APPROVED** [review]

**Comment by jdevcs** [comment]
lgtm, plus we should add simplest blackbox tests as well https://github.com/ChainSafe/web3.js/issues/5261 as part of new PR

**Review by avkos — APPROVED** [review]

**Pull request merged**